zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1865,17 +1934,24 @@ public Object call() throws Exception {
    *         exist.
    */
   public List<String> watchForChilds(final String path) {
+    return watchForChilds(path, false);
+  }
+
+  private List<String> watchForChilds(final String path, boolean skipWatchingNodeNotExist) {
     if (_zookeeperEventThread != null && Thread.currentThread() == _zookeeperEventThread) {
       throw new IllegalArgumentException("Must not be done in the zookeeper event thread.");
     }
     return retryUntilConnected(new Callable<List<String>>() {
       @Override
       public List<String> call() throws Exception {
-        exists(path, true);
+        if (!skipWatchingNodeNotExist) {
```

Comment:
If skipWatchingNodeNotExist, then we don't watch the root path? This change the behavior, right?

Reply:
If skipWatchingNodeNotExist, we don't add watch to parent path. That is exactly what we want.

We can think it this way. 
1/ skipWatchingNodeNotExist true, if the parent path is not there. We will not install watch.
2/ skipWatchingNodeNotExist false, namely original behavior, if the parent path is not there, we will install exist data watch.

We discussed this in the meeting. 

The second case is actually needed for current state callbackhandler registration. LiveInstance change would cause current state callbackhandler to be added to ZkHelixManager. And at the time, it can often be the case the parent node of current state is not yet created. Note, this is the INIT case.

Later CALLBACK case, we should expect the current state parent path be there. Otherwise, we should not register the watch to Zk server (or we will leak a path). That is why we need the first case.

Reply:
This is intentional. 

if skipWatchingNodeNoExist is true, we skip install **"data"** watch using exists(). Thus, if the directory (parent path) is not created, it won't leak a "data" watch in server. 
Note, the getChildren() also installs **"child"** watch if the parent path exists(). Otherwise, it would return null. 

Note **child** watch and **data** watch are two different watch from ZK server perspective, stored in different data structure.  **Data** watch can be installed to path not exist yet, **child** watch can't.

If the parent path exists, we only need child watch to function correctly. skipWatchingNodeNotExist is true in this case. This is also the case when CallbackHandler has CALLBACK notification.

If we want to wait till paren path to be created, we set skipWatchingNodeNotExist to be false. This is the case when CallbackHandler has INIT notification.

Reply:
> 1/ skipWatchingNodeNotExist true, if the parent path is not there. We will not install watch.

What you are doing is, if skipWatchingNodeNotExist is true, then not install the watch no matter the parent path exists or not. Please double check.

Reply:
@kaisun2000 I dont think that is a good idea. In this case, it only handle the leakage of data change but not child change. I think it is better to have consistent behavior: if node does not exist and skipWatchingNodeNotExist is true, no data change and child change watcher install. Otherwise, install both.

Reply:
@jiajunwang, the code is like this:
```
     public List<String> call() throws Exception {
        if (!skipWatchingNodeNotExist) {
          exists(path, true);
        }
        try {
          return getChildren(path, true);
        } catch (ZkNoNodeException e) {
          // ignore, the "exists" watch will listen for the parent node to appear
          LOG.info("watchForChilds path not existing:" + path + " skipWatchingNodeNoteExist:" + skipWatchingNodeNotExist);
        }
        return null;
      }
```

 if skipWatchingNodeNotExist is true, they call getChildren(). 

- If the parent path exists, it would install child change. 

- Otherwise, if parent path not exists, it would not  install parent path. Instead, it would return null as child list. This would eventually cause the CallbackHandler's handleChildren() code path `_manager.removeListener(_propertyKey, _listener);`. This path would remove CallbackHandler.

Reply:
@dasahcc, this part is kind of complicated. Seems you are concerned that there are some insistent state. But fortunately, we can reason this way, which covers all the cases.

"if node does not exist and skipWatchingNodeNotExist is true", in some case, we do need to install data watch if we expect the parent node is to be created. 

Let us look at the life cycle of CallbackHandler (CB) to see all the cases.

- INIT of CallbackHandler when parent node note existing. CB first installs data(exists) watch to this parent node. Then when the node is created, CB got notification of creation. Then CB will read all of the children (installing child watch at this time) and install data watch for each children. 

- INIT of CallbackHandler when parent node exists. CB still first installs data(exists) watch to this parent node. And the use getData() to get the list of children and install data watch for each children. 

Not the above two code path are the same. They are started from CB's subscribeForChanges() --> subscribeChildChange(INIT) --> _zkClient.subscribeChildChanges( skipWatchingNodeNotExist as false) --> watchForChild(skipWatchingNodeNotExist as false)

This would trigger both exists() and getData(). This is exactly when we want, because in the first case, you need exists() if the parent path does not exists as explained before. 

Then you next question can be why for the second case I need exists() if the path does exists? The is because later, if the paren path is removed, you want to remove the CB from ZkHelixManager too. With data watch, you are guaranteed a notification of handleChildChange(). Here the `_manager.removeListener(_propertyKey, _listener);` code path would be trigger. This is exactly what happens when current state parent node moves due to new session established in participant code. 

So above discussion addressed the basic life cycle of CallbackHandler. 

The left case is about what happens when normal notification of CALLBACK type.
This is triggered by adding or deleting a current state node under the parent path. At this time, the code path is: 

subscribeForChanges() --> subscribeChildChange(CALLBACK) --> _zkClient.subscribeChildChanges( skipWatchingNodeNotExist as true) --> watchForChild(skipWatchingNodeNotExist as true)

Here, watchForChild will not install data watch for parent node, only use getData to install child watch and return list of children. Then CallbackHandler would install data watch for each children. 

So here, if you think about it, there is no in-consistant case of installing part of the parent child watch, but not child data watch. 

Note, we should not install data watch to parent node here at CALLBACK stage for the reason that the removing of parent node is async to this whole process. That can be the source of leaking once the INIT phase passed, as I explained in the doc.

The above discussion exhausts all the cases of CallbackHandler life cycle. Hope this would give people a clear idea of how we reason it.

Reply:
Synced with @kaisun2000 offline. It should be potential leakage but minimum risk.

My final comment: 

Here that you introduced the new flag to change the logic. But two types of watchers have different behavior for that flag. This could be confusing.  If you think it is necessary to keep two different behavior, please add more comments of the code. That could lead code reader very confusing.

Reply:
Thx, @dasahcc, let me summarize our the gist of our discussion  -- async removal and exists() compatibility and life cycle of callbackhandler here.

Indeed this is an arcane area hard for people to understand.

Reply:
This is good and intelligent discussion. I add the comment and would resolve this issue. Let me know if other wise.,